verification/cbmc/proofs/sign_header/sign_header_harness.c

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/cryptosdk/private/cipher.h>
+#include <aws/cryptosdk/private/session.h>
+#include <aws/cryptosdk/session.h>
+#include <cbmc_invariants.h>
+#include <make_common_data_structures.h>
+#include <proof_helpers/make_common_data_structures.h>
+
+/**
+ * A generator function as described in the comment in aws_cryptosdk_hash_elems_array_init_stub.c.
+ * Also see line 33 of the Makefile.
+ */
+void array_list_item_generator(struct aws_array_list *elems) {
+    assert(elems->item_size == sizeof(struct aws_hash_element));
+    for (size_t index = 0; index < elems->length; ++index) {
+        struct aws_hash_element *val = (struct aws_hash_element *)((uint8_t *)elems->data + (elems->item_size * index));
+        /* Due to the checks in aws_cryptosdk_enc_ctx_size, no string can have a length > UINT16_MAX. */
+        struct aws_string *key = ensure_string_is_allocated_nondet_length();
+        __CPROVER_assume(aws_string_is_valid(key));
+        /* Due to the cast to uint16, the entire size of the enc_ctx must be less than < UINT16_MAX
+         * This is a simple way to ensure this without a call to enc_ctx_size. */
+        __CPROVER_assume(key->len <= UINT8_MAX);
+        val->key                 = key;
+        struct aws_string *value = ensure_string_is_allocated_nondet_length();
+        __CPROVER_assume(aws_string_is_valid(value));
+        __CPROVER_assume(value->len <= UINT8_MAX);
+        val->value = value;
+    }
+}
+
+void sign_header_harness() {
+    /* Nondet Input */
+    struct aws_cryptosdk_session *session = malloc(sizeof(*session));
+
+    /* Assumptions */
+    __CPROVER_assume(session != NULL);
+
+    enum aws_cryptosdk_alg_id alg_id;
+    struct aws_cryptosdk_alg_properties *props = aws_cryptosdk_alg_props(alg_id);
+    __CPROVER_assume(aws_cryptosdk_alg_properties_is_valid(props));
+    session->alg_props = props;
+
+    struct aws_cryptosdk_hdr *hdr = ensure_nondet_hdr_has_allocated_members(MAX_TABLE_SIZE);
+    __CPROVER_assume(aws_cryptosdk_hdr_members_are_bounded(hdr, MAX_EDK_LIST_ITEMS, MAX_BUFFER_SIZE));
+    __CPROVER_assume(IMPLIES(hdr != NULL, aws_byte_buf_is_bounded(&hdr->iv, session->alg_props->iv_len)));
+    __CPROVER_assume(IMPLIES(hdr != NULL, aws_byte_buf_is_bounded(&hdr->auth_tag, session->alg_props->tag_len)));
```

Comment:
Without being familiar with `aws_cryptosdk_hdr_members_are_bounded`, these lines are confusing to me - if line 58 assumes that the all header members are bounded, why are lines 59-60 necessary to assume that some particular header members are bounded?

Reply:
So this has been partially moved into hdr_setup, but basically the generic header setup assumes that all the byte_bufs fields of hdr are bounded by a MAX_BUFFER_SIZE. This MAX_BUFFER_SIZE comes from Makefile.aws_byte_buf and is set to 1028. However, for the iv and auth_tag having a buffer that large will actually result in overflow. Therefore I add the additional assumptions that they are bounded by session->alg_props->iv_len and session->alg_props->tag_len respectively (in fact I think they should be equal?). Is that what you would expect for the iv and auth_tag buffers?

Reply:
I see, I think you're right - the buffer lengths are indeed equal-to/bound-by the `alg_props->{iv,tag}_len` values. I had missed the different bounds in my original comment.